File: packages/core/src/city.ts
/**
 * City — canonical city record with geonameId for deduplication.
 *
 * Multiple language variants of the same city (e.g. "万象", "Vientiane",
 * "ວຽງຈັນ") share the same geonameId. The seed loader dedupes by geonameId
 * so only one row per real city enters the database.
 */

export interface City {
  /** GeoNames unique identifier. Stable across language variants. */
  geonameId: number;
  /** Name in the local/native script of the country. */
  nameLocal: string;
  /** Name in the system/display language (romanized or English). */
  nameSystem: string;
  /** WGS-84 latitude. */
  lat: number;
  /** WGS-84 longitude. */
  lon: number;
  /** ISO 3166-1 alpha-2 country code, e.g. "TH", "LA", "VN". */
  countryCode: string;
}

/**
 * Deduplicate an array of City records by geonameId.
 * First occurrence wins (inner-first semantics, same as Overpass dedupe).
 */
export function dedupeByGeonameId(cities: City[]): City[] {
  const seen = new Set<number>();
  return cities.filter((c) => {
    if (seen.has(c.geonameId)) return false;
    seen.add(c.geonameId);
    return true;
  });
}

File: apps/ios/SoloCompass/Models/City.swift
import Foundation

/// Canonical city record mirroring packages/core/src/city.ts.
/// geonameId is the stable deduplication key across language variants.
public struct City: Codable, Identifiable, Hashable {
    /// GeoNames unique identifier. Stable across language variants.
    public let geonameId: Int
    /// Name in the local/native script of the country.
    public let nameLocal: String
    /// Name in the system/display language (romanized or English).
    public let nameSystem: String
    /// WGS-84 latitude.
    public let lat: Double
    /// WGS-84 longitude.
    public let lon: Double
    /// ISO 3166-1 alpha-2 country code, e.g. "TH", "LA", "VN".
    public let countryCode: String

    public var id: Int { geonameId }

    public init(
        geonameId: Int,
        nameLocal: String,
        nameSystem: String,
        lat: Double,
        lon: Double,
        countryCode: String
    ) {
        self.geonameId = geonameId
        self.nameLocal = nameLocal
        self.nameSystem = nameSystem
        self.lat = lat
        self.lon = lon
        self.countryCode = countryCode
    }
}

// MARK: - Dedupe

extension Array where Element == City {
    /// Returns a new array with duplicates by geonameId removed.
    /// First occurrence wins.
    public func deduplicatedByGeonameId() -> [City] {
        var seen = Set<Int>()
        return filter { seen.insert($0.geonameId).inserted }
    }
}
